content/blog/coreboot-on-talos2.md

```diff
@@ -0,0 +1,307 @@
+    ```
+
+Answer yes to the prompt and wait for the process to finish. After that, start
+the platform and begin [configuring Heads](https://osresearch.net/Configuring-Keys/).
+
+## PNOR emulation
+
+Flash device can be emulated by BMC, which is something we were often using for
+development and testing. This saves a lot of time which would be spent flashing,
+as well as reduces the wear of flash device.
+
+However, this still requires System Package v2.00, and if this is different than
+what real flash holds, SEEPROM will have to be updated when switching between
+physical and emulated image. Also, this approach doesn't survive BMC reboots and
+power losses. BMC doesn't have enough space to keep full PNOR image in
+non-volatile memory, so `tmpfs` must be used for emulation. Don't try to put
+more than one image in `tmpfs` or BMC **will** run out of RAM, which most likely
+will require manual power cycle.
+
+To start, obtain full flash image, either by downloading it from
+[RaptorCS release page](https://wiki.raptorcs.com/wiki/Talos_II/Firmware) or
+reading from existing image on BMC with:
+
+```
+pflash -r /tmp/talos.pnor
+```
+
+After that, you can "flash" the partitions mentioned earlier by adding
+additional parameters to use the file instead of physical flash:
+
+```
+pflash -f -P <partition> -p <image_file> -F /tmp/talos.pnor
+```
+
+Change `<partition>` to one of `HBB`, `HBI`, `BOOTKERNEL` and `<image_file>` to
+`/tmp/bootblock.signed.ecc`, `/tmp/coreboot.rom.signed.ecc` or
+`/tmp/zImage.bundled`, respectively.
+
+To tell BMC to present the contents of this file as flash, run:
+
+```
+mboxctl --backend file:/tmp/talos.pnor
+```
+
+Sometimes this command fails with a timeout, in that case run it again until it
+succeeds.
+
+> We've noticed that sometimes, despite no error message printed, physical flash
+> was used anyway. It is easy to spot when one copy has Hostboot and the other
+> has coreboot, but it can be missed when both images have different versions of
+> coreboot. It caused us few hours of unnecessary debugging of issues that were
+> already fixed...
+
+With the file now mounted, platform can be started. Host firmware and OS
+shouldn't be able to tell the difference, except for different reported erase
+block size and maybe different access times.
+
+To get back to original flash, run:
+
+```
+mboxctl --backend vpnor
+```
+
+It will report an error (`Failed to post message: Connection timed out`), but
+will revert to physical device nonetheless. This can be confirmed by checking
+the output of `mboxctl --lpc-state`:
+
+```
+root@talos:~# mboxctl --lpc-state
+LPC Bus Maps: Flash Device
+```
+
+Since the file is now the full image with coreboot (and optionally Heads), it
+can be simply written to flash, should you choose to accept it:
+
+```
+pflash -E -p /tmp/talos.pnor
+```
+
+## Noticeable differences between Hostboot and coreboot
+
+For those wandering why we even started this project, here are some of the
```

Comment:
```suggestion
For those wondering why we even started this project, here are some of the
```

@krystian-hebel applying spellcheck wouldn't hurt

Reply:
Which is what I have always enabled in my editor, but this isn't spelling error. Thanks for catching it.
https://github.com/3mdeb/openpower-foundation-website/compare/cdb1949a6559800e6bb13cbea70060348116da21..6a92856303bf2f67b7522a0c5711f5ac5322537e

Reply:
I understand. Grammarly could possibly catch it.